JIT: Optimize matching of utf8 segments

bjorng opened the pull request:
Currently, matching a Unicode code point as UTF-8 in the binary syntax is done by calling a helper function written in C.

Optimize this by inlining all code for handling the ASCII subset of Unicode (code points less than 128). Handle other code points by calling a helper fragment implemented in native code. Calling a fragment has slightly less overhead than calling a function implemented in C.

Co-authored-by: John Högberg <john@erlang.org>

michalmuskala commented:
I wonder if it would make sense to explore using a DFA-based approach for loading utf8 segments. See description in: https://bjoern.hoehrmann.de/utf-8/decoder/dfa/

AFAIK today implementing this in Erlang is actually faster than using `/utf8` patterns (this is what [Cowboy does](https://github.com/ninenines/cowlib/blob/6deddc7d3360aa0f50eb2375cc0226157185c472/src/cow_ws.erl#L581) to verify strings are utf8) - it's possible this PR will change this, but I think it might be interesting to explore this table-based approach. It also has the advantage of decoding all codepoints, not only those < 128.

bjorng commented on erts/emulator/beam/jit/arm/instr_bs.cpp:
```diff
@@ -1028,18 +1028,195 @@ void BeamModuleAssembler::emit_i_bs_put_utf8(const ArgLabel &Fail,
+    a.csetm(TMP2, imm(arm::CondCode::kNE));
+    a.add(TMP1, TMP1, TMP2);
+
+    /* Now isolate the y bits and compare to zero. */
+    a.lsr(TMP1, bitdata, TMP1);
+    a.cmp(TMP1, 0);
+    a.csel(byte_count, byte_count, ZERO, imm(arm::CondCode::kNE));
+
+    /* Byte count must be 2, 3, or 4. */
+    a.sub(TMP1, byte_count, imm(2));
+    a.cmp(TMP1, imm(2));
+    a.b_hi(error);
+
+    /* Now we know that the match has succeeded. Update the position
+     * in the match state. */
+    a.add(bin_position, bin_position, byte_count, arm::lsl(3));
+    a.str(bin_position, arm::Mem(match_state, position_offset));
+
+    /* Tag the integer value. */
+    mov_imm(ARG1, _TAG_IMMED1_SMALL);
+    a.orr(ARG1, ARG1, bitdata, arm::lsl(_TAG_IMMED1_SIZE));
+
+    a.ret(a64::x30);
```
Yes, that works.

bjorng commented on erts/emulator/beam/jit/arm/instr_bs.cpp:
```diff
@@ -1028,18 +1028,195 @@ void BeamModuleAssembler::emit_i_bs_put_utf8(const ArgLabel &Fail,
+    a.orr(bitdata, bitdata, TMP3, arm::lsl(18));
+
+    /* Check for too large code point. */
+    cmp(bitdata, 0x10FFFF);
+    a.csel(byte_count, byte_count, ZERO, imm(arm::CondCode::kLS));
+
+    /* Check for the illegal range 16#D800 - 16#DFFF. */
+    a.lsr(TMP1, bitdata, imm(11));
+    a.cmp(TMP1, imm(0xD800 >> 11));
+    a.csel(byte_count, byte_count, ZERO, imm(arm::CondCode::kNE));
```
That works!

bjorng commented on erts/emulator/beam/jit/arm/instr_bs.cpp:
```diff
@@ -1028,18 +1028,195 @@ void BeamModuleAssembler::emit_i_bs_put_utf8(const ArgLabel &Fail,
+
+    const arm::Gp byte_count = ARG4;
+
+    const arm::Gp shift = TMP2;
+    const arm::Gp control_mask = TMP4;
+    const arm::Gp error_mask = TMP5;
```
I agree. Fixed.

bjorng commented on erts/emulator/beam/jit/arm/instr_bs.cpp:
```diff
@@ -1028,18 +1028,195 @@ void BeamModuleAssembler::emit_i_bs_put_utf8(const ArgLabel &Fail,
+    fragment_call(ga->get_bs_get_utf8_shared());
+    a.b(check);
 
+    /*
+     * Handle the case that there are not 4 bytes available in the binary.
+     */
+
+    a.bind(fallback);
+    emit_enter_runtime();
+    runtime_call<1>(erts_bs_get_utf8);
     emit_leave_runtime();
```
Yes, that is probably a good idea. I'll try it tomorrow.

bjorng replied:
It did bother me that I had to have two fallback routines.

bjorng replied:
I have pushed a `fixup` commit that implements that idea for AArch64.

bjorng replied:
Now done for x86 as well.

bjorng commented:
@michalmuskala It seems that his implementation does not reject overlong UTF-8 sequences or the invalid characters in the range 16#D800 through 16#DFFF, which we do.

jhogberg commented on erts/emulator/beam/jit/arm/instr_bs.cpp:
```diff
@@ -1028,18 +1028,192 @@ void BeamModuleAssembler::emit_i_bs_put_utf8(const ArgLabel &Fail,
+    a.orr(bitdata, bitdata, TMP3, arm::lsl(18));
+
+    /* Check for too large code point. */
+    cmp(bitdata, 0x10FFFF);
```
We ought to be able to get rid of another `CSEL` by making this a (register-variant) `CCMP` too.

bjorng replied:
Done.

michalmuskala commented:
> @michalmuskala It seems that his implementation does not reject overlong UTF-8 sequences or the invalid characters in the range 16#D800 through 16#DFFF, which we do.

I wrote a [simple program](https://gist.github.com/michalmuskala/77dfecdb57c82e4d7a51c0c4df815be6) to check how those are handled, this is the output:
```
Checking surrogate: "\xD8\x01"
The string is not well-formed
Checking surrogate: "\xDF\xFE"
The string is not well-formed
Checking regular A: "\x41"
U+0041
Checking overlong A: "\xC0\xA0"
The string is not well-formed
```

I believe surrogates and overlongs are correctly rejected by the algorithm

bjorng commented:
@jhogberg has implemented an improved version of Björn Hörmann's DFA algorithm. It turns out that the speed is about the same as for this PR, with perhaps a very slight improvement for some input data.

As John's improvements involved building the lookup table in a different way, shaving off some instructions in using the table (resulting in a speed up of about 10 percent), that implementation turned out to be quite complex and "magic", not at all easy to understand.

Therefore, we have decided to use this PR. I have now fixed two bugs and added it to our daily builds, and we plan to merge it next week if everything goes well.

josevalim commented:
This library may also be of interest: https://github.com/simdutf/simdutf - it contains papers references at the bottom.

jhogberg commented:
Yes, we've looked at various ways to leverage SIMD, but unfortunately it's not very effective for extracting one codepoint at a time, which is how most code looks today. :-|

mtrudel commented:
> Yes, we've looked at various ways to leverage SIMD, but unfortunately it's not very effective for extracting one codepoint at a time, which is how most code looks today. :-|

There remains a number of use cases beyond matching that would benefit from SIMD, namely UTF-8 validation/codepoint counting. Indeed, the Cowboy routines referenced above exist solely for the purpose of validating if a binary is well-formed UTF-8 (this being, for example, a necessary validation for WebSocket frames per RFC6455). Currently, Elixir's standard library validates UTF-8 by recursively [matching a single UTF-8 code point](https://github.com/elixir-lang/elixir/blob/a64d42f5d3cb6c32752af9d3312897e8cd5bb7ec/lib/elixir/lib/string.ex#L1715), and ends up being an extremely hot code path for many applications. 

I'm currently [investigating](https://github.com/elixir-lang/elixir/pull/12354) some improvements to Elixir's UTF-8 validation routine for all-ASCII matching that have emerged with OTP26 (though not from the work on this PR). However, a natively supported SIMD approach such as is provided by simdutf and which provided standalone validation and code point counting routines would be an *extremely* useful adjunct to the (also very welcome!) improvements to matching included here. 

Speaking as a 'long-time user, first-time contributor', is there a preferred way to surface such a suggestion?

bjorng commented:
@mtrudel 

> Speaking as a 'long-time user, first-time contributor', is there a preferred way to surface such a suggestion?

I suggest you first discuss your ideas in the [erlang forums](https://erlangforums.com). When you are ready to make a pull request, be sure to read our guidlines for [Contributing to Erlang/OTP](https://github.com/erlang/otp/blob/master/CONTRIBUTING.md).
